### DISCUSSION TITLE
overlapping stacks

### DISCUSSION THREAD

**Discussion by allan-bonadio** [emscripten-core/emscripten#18920] [Q&A] [answered]
I've got a main thread, and two workers, sharing shared memory, the usual 16 megs.  Their stacks all seem to overlap with each other in the beginning 64kiB of the memory, which I guess is the default.  This doesn't sound healthy to me, aren't they all just going to overwrite each other?

I've tried to put them in different pages like this:
	Module.asm.__set_stack_limits((serial+2) * 65536, (serial+1) * 65536);
where serial is an integer telling which worker number, 0, 0, ...
but emscripten_stack_get_base() and the others always return the same numbers.  And, everything just crashes.  How am I supposed to make room for the different stacks?

**Comment by sbc100** [comment, marked as answer]
Are you using `pthread_create` to create your workers/threads?  The emscripten runtime takes care of allocating a different stack for each thread during the pthread creation process.

**Reply by allan-bonadio** [reply]
no, I'vew been trying to use web-workers.  Waste of time.  I'm trying pthreads now, but it's huge.  Do you have a schedule for when web-workers will be functional?

**Reply by sbc100** [reply]
When you say web-workers do you mean wasm-workers?   Web Workers are the thing that underpins both pthreads and wasm-workers.  I admit the terminology can be confusing.

If you mean wasm-workers then those are already supported.  If you have specific issues using wasm-workers then perhaps @juj can help.   They should not have overlapping stacks, that is for sure.

If you really do mean Web Workers and, for example, you are trying to create new ones by calling `new Worker` yourself, then that is *not* something that is supported by emscripten, or at least you would have to jump through all kind of hoops and basically re-implement wasm-workers/pthreads yourself.

**Reply by allan-bonadio** [reply]
sorry, yeah, wasm-workers.  [Peter Piper Principle](https://en.wikipedia.org/wiki/Peter_Piper#Peter_Piper_Principle).  At first glance, seems it should be easy: just new Worker() then start calling my usual C++ methods.  Never realized how complex it could be till I started reading the generated JS sourcecode.  Yeah, I'll contact @juj in the future.  I'm off trying pthreads.  

Thanks for a great SDK!

**Reply by sbc100** [reply]
To create wasm workers you need to use the wasm workers API to do that.  e.g `emscripten_malloc_wasm_worker` or `emscripten_create_wasm_worker`.    You cannot, IIUC, just call `new Worker` yourself and have that just work.
